File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/access/AuthSessionProvider.kt
package com.reguerta.user.domain.access

enum class AuthSignInFailureReason {
    INVALID_EMAIL,
    INVALID_CREDENTIALS,
    EMAIL_ALREADY_IN_USE,
    WEAK_PASSWORD,
    USER_NOT_FOUND,
    USER_DISABLED,
    TOO_MANY_REQUESTS,
    NETWORK,
    UNKNOWN,
}

sealed interface AuthSignInResult {
    data class Success(val principal: AuthPrincipal) : AuthSignInResult

    data class Failure(val reason: AuthSignInFailureReason) : AuthSignInResult
}

sealed interface AuthPasswordResetResult {
    data object Success : AuthPasswordResetResult

    data class Failure(val reason: AuthSignInFailureReason) : AuthPasswordResetResult
}

sealed interface AuthSessionRefreshResult {
    data object NoSession : AuthSessionRefreshResult

    data class Active(val principal: AuthPrincipal) : AuthSessionRefreshResult

    data object Expired : AuthSessionRefreshResult
}

interface AuthSessionProvider {
    suspend fun signIn(email: String, password: String): AuthSignInResult

    suspend fun signUp(email: String, password: String): AuthSignInResult

    suspend fun sendPasswordReset(email: String): AuthPasswordResetResult

    suspend fun refreshCurrentSession(): AuthSessionRefreshResult

    fun signOut()
}

File: ios/Reguerta/Reguerta/Domain/Access/AuthSessionProvider.swift
import Foundation

enum AuthSignInFailureReason: Equatable, Sendable {
    case invalidEmail
    case invalidCredentials
    case emailAlreadyInUse
    case weakPassword
    case userNotFound
    case userDisabled
    case tooManyRequests
    case network
    case unknown
}

enum AuthSignInResult: Equatable, Sendable {
    case success(AuthPrincipal)
    case emailVerificationRequired(email: String, verificationResent: Bool, signedOut: Bool)
    case failure(AuthSignInFailureReason)
}

enum AuthSignUpResult: Equatable, Sendable {
    case verificationRequired(email: String, verificationSent: Bool, signedOut: Bool)
    case failure(AuthSignInFailureReason)
}

enum AuthPasswordResetResult: Equatable, Sendable {
    case success
    case failure(AuthSignInFailureReason)
}

enum AuthSessionRefreshResult: Equatable, Sendable {
    case noSession
    case active(AuthPrincipal)
    case emailVerificationRequired(email: String)
    case failure(AuthSignInFailureReason)
    case expired
}

@MainActor
protocol FirebaseIDTokenProviding {
    func validIDToken(forcingRefresh: Bool) async throws -> String
}

@MainActor
protocol AuthSessionProvider: FirebaseIDTokenProviding {
    func signIn(email: String, password: String) async -> AuthSignInResult
    func signUp(email: String, password: String) async -> AuthSignUpResult
    func sendPasswordReset(email: String) async -> AuthPasswordResetResult
    func sendCurrentUserEmailVerification() async -> Bool
    func refreshCurrentSession() async -> AuthSessionRefreshResult
    @discardableResult func signOut() -> Bool
}
